File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/servers/library/items/media/files/properties/uri/BestMatchUriProvider.java
package com.lasthopesoftware.bluewater.servers.library.items.media.files.properties.uri;

import android.Manifest;
import android.content.Context;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.support.v4.content.ContextCompat;

import com.lasthopesoftware.bluewater.servers.connection.ConnectionProvider;
import com.lasthopesoftware.bluewater.servers.library.items.media.files.IFile;
import com.lasthopesoftware.bluewater.servers.library.repository.Library;

import java.io.IOException;

/**
 * Created by david on 7/24/15.
 * Will get the best URI for access speed.
 */
public class BestMatchUriProvider extends AbstractFileUriProvider {
	private final Context context;
	private final Library library;
	private final ConnectionProvider connectionProvider;

	public BestMatchUriProvider(Context context, ConnectionProvider connectionProvider, Library library, IFile file) {
		super(file);

		this.context = context;
		this.library = library;
		this.connectionProvider = connectionProvider;
	}

	@Override
	public Uri getFileUri() throws IOException {
		final StoredFileUriProvider storedFileUriProvider = new StoredFileUriProvider(context, library, getFile());
		Uri fileUri = storedFileUriProvider.getFileUri();
		if (fileUri != null)
			return fileUri;

		if (library.isUsingExistingFiles() && ContextCompat.checkSelfPermission(context, Manifest.permission.READ_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED) {
			final MediaFileUriProvider mediaFileUriProvider = new MediaFileUriProvider(context, connectionProvider, getFile());
			fileUri = mediaFileUriProvider.getFileUri();
			if (fileUri != null)
				return fileUri;
		}

		final RemoteFileUriProvider remoteFileUriProvider = new RemoteFileUriProvider(connectionProvider, getFile());
		return remoteFileUriProvider.getFileUri();
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/browsing/files/uri/BestMatchUriProvider.kt
package com.lasthopesoftware.bluewater.client.browsing.files.uri

import android.net.Uri
import com.lasthopesoftware.bluewater.client.browsing.files.ServiceFile
import com.lasthopesoftware.bluewater.client.browsing.library.access.ProvideLibraries
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.client.stored.library.items.files.external.MediaFileUriProvider
import com.lasthopesoftware.bluewater.client.stored.library.items.files.uri.StoredFileUriProvider
import com.lasthopesoftware.promises.extensions.toPromise
import com.namehillsoftware.handoff.promises.Promise

class BestMatchUriProvider(
    private val libraryProvider: ProvideLibraries,
    private val storedFileUriProvider: StoredFileUriProvider,
	private val mediaFileUriProvider: MediaFileUriProvider,
    private val remoteFileUriProvider: ProvideFileUrisForLibrary,
) : ProvideFileUrisForLibrary {
    override fun promiseUri(libraryId: LibraryId, serviceFile: ServiceFile): Promise<Uri?> =
        storedFileUriProvider
            .promiseUri(libraryId, serviceFile)
            .eventually { storedFileUri ->
				storedFileUri
					?.toPromise()
					?: libraryProvider
						.promiseLibrary(libraryId)
						.eventually { library ->
							when {
								library == null -> Promise.empty()
								!library.isUsingExistingFiles -> remoteFileUriProvider.promiseUri(libraryId, serviceFile)
								else -> mediaFileUriProvider
									.promiseUri(libraryId, serviceFile)
									.eventually { mediaFileUri ->
										mediaFileUri?.toPromise() ?: remoteFileUriProvider.promiseUri(libraryId, serviceFile)
									}
							}
						}
			}
}
